Common/Data/LeanDataWriter.cs

```diff
@@ -172,340 +214,178 @@ public void DownloadAndSave(IBrokerage brokerage, List<Symbol> symbols, DateTime
             }
 
-            // Loop through the SortedDictionary and write to file contents
-            foreach (var kvp in rows)
+            if (_appendToZips)
             {
-                // Build the line and append it to the file
-                lines.Add(kvp.Value);
-            }
+                var bytes = Encoding.UTF8.GetBytes(string.Join("\n", rows.Values));
+                _dataCacheProvider.Store($"{filePath}#{entryName}", bytes);
 
-            // Write the file contents
-            if (lines.Count > 0)
+                Log.Trace($"LeanDataWriter.Write(): Appended: {filePath} @ {entryName}");
+            }
+            else
             {
-                WriteFile(outputFile, lines, lastTime);
+                // Write out this data string to a zip file
+                var tempFilePath = filePath + ".tmp";
+                Compression.ZipData(tempFilePath, entryName, rows.Values);
+
+                // Move temp file to the final destination with the appropriate name
+                File.Move(tempFilePath, filePath);
+                Log.Trace("LeanDataWriter.Write(): Created: " + filePath);
             }
         }
 
         /// <summary>
-        /// Loads an existing hourly or daily Lean zip file into a SortedDictionary
+        /// Get the output zip file
         /// </summary>
-        private static SortedDictionary<DateTime, string> LoadHourlyOrDailyFile(string fileName)
+        /// <param name="baseDirectory">Base output directory for the zip file</param>
+        /// <param name="time">Date/time for the data we're writing</param>
+        /// <returns>The full path to the output zip file</returns>
+        private string GetZipOutputFileName(string baseDirectory, DateTime time)
         {
-            var rows = new SortedDictionary<DateTime, string>();
+            return LeanData.GenerateZipFilePath(baseDirectory, _symbol, time, _resolution, _tickType);
+        }
 
-            using (var zip = ZipFile.Read(fileName))
-            {
-                using (var stream = new MemoryStream())
-                {
-                    zip[0].Extract(stream);
-                    stream.Seek(0, SeekOrigin.Begin);
+    }
 
-                    using (var reader = new StreamReader(stream))
-                    {
-                        string line;
-                        while ((line = reader.ReadLine()) != null)
-                        {
-                            var time = Parse.DateTimeExact(line.Substring(0, DateFormat.TwelveCharacter.Length), DateFormat.TwelveCharacter);
-                            rows[time] = line;
-                        }
-                    }
-                }
-            }
+    /// <summary>
+    /// Simple data cache provider, writes and reads directly from disk
+    /// Used as default for <see cref="LeanDataWriter"/>
+    /// </summary>
+    public class DiskDataCacheProvider : IDataCacheProvider
```

Comment:
Cool! let's move this guy to his own class file plus  a couple of tests would be great! 👍